File: backend/supabase/functions/_shared/scriptService.ts
import { sql } from "./db.ts";
import { d } from "./decimal.ts";
import { symbolExistsOnNse } from "./marketService.ts";

export type Script = {
  id: number;
  name: string;
  up_percent: string;
  down_percent: string;
  total_invest_amount: string;
  invest_slots: number;
  max_buy: number;
  active: boolean;
  created_at: Date;
};

export type ScriptInput = {
  name: string;
  up_percent: number;
  down_percent: number;
  total_invest_amount: number;
  invest_slots: number;
  max_buy: number;
  active: boolean;
};

export class ScriptNotFoundOnNseError extends Error {}

export function normalizeScriptName(name: string): string {
  return name.trim().toUpperCase();
}

export function listScripts(): Promise<Script[]> {
  return sql<Script[]>`select * from scripts order by created_at desc`;
}

export async function getScript(scriptId: number): Promise<Script | null> {
  const rows = await sql<Script[]>`select * from scripts where id = ${scriptId}`;
  return rows[0] ?? null;
}

export async function createScript(payload: ScriptInput): Promise<Script> {
  const name = normalizeScriptName(payload.name);
  if (!(await symbolExistsOnNse(name))) {
    throw new ScriptNotFoundOnNseError();
  }
  const [row] = await sql<Script[]>`
    insert into scripts (name, up_percent, down_percent, total_invest_amount, invest_slots, max_buy, active)
    values (
      ${name},
      ${d(payload.up_percent).toString()},
      ${d(payload.down_percent).toString()},
      ${d(payload.total_invest_amount).toString()},
      ${payload.invest_slots},
      ${payload.max_buy},
      ${payload.active}
    )
    returning *
  `;
  return row;
}

export async function updateScript(scriptId: number, payload: ScriptInput): Promise<Script | null> {
  const existing = await getScript(scriptId);
  if (!existing) return null;
  const name = normalizeScriptName(payload.name);
  if (name !== existing.name && !(await symbolExistsOnNse(name))) {
    throw new ScriptNotFoundOnNseError();
  }
  const [row] = await sql<Script[]>`
    update scripts set
      name = ${name},
      up_percent = ${d(payload.up_percent).toString()},
      down_percent = ${d(payload.down_percent).toString()},
      total_invest_amount = ${d(payload.total_invest_amount).toString()},
      invest_slots = ${payload.invest_slots},
      max_buy = ${payload.max_buy},
      active = ${payload.active}
    where id = ${scriptId}
    returning *
  `;
  return row;
}

export async function deleteScript(scriptId: number): Promise<boolean> {
  const rows = await sql`delete from scripts where id = ${scriptId} returning id`;
  return rows.length > 0;
}

File: backend/app/services/script_service.py
from decimal import Decimal

from sqlalchemy.orm import Session

from app.models.script_model import Script
from app.schemas.script_schema import ScriptCreate, ScriptUpdate
from app.services.market_service import symbol_exists_on_nse


def _d(value: object) -> Decimal:
    dec = value if isinstance(value, Decimal) else Decimal(str(value))
    return dec.quantize(Decimal("0.01"))


def list_scripts(db: Session) -> list[Script]:
    return db.query(Script).order_by(Script.created_at.desc()).all()


def get_script(db: Session, script_id: int) -> Script | None:
    return db.query(Script).filter(Script.id == script_id).first()


def create_script(db: Session, payload: ScriptCreate) -> Script:
    if not symbol_exists_on_nse(payload.name):
        raise ValueError("SCRIPT_NOT_FOUND")
    row = Script(
        name=payload.name,
        up_percent=_d(payload.up_percent),
        down_percent=_d(payload.down_percent),
        total_invest_amount=_d(payload.total_invest_amount),
        invest_slots=payload.invest_slots,
        max_buy=payload.max_buy,
        active=payload.active,
    )
    db.add(row)
    db.commit()
    db.refresh(row)
    return row


def update_script(db: Session, script_id: int, payload: ScriptUpdate) -> Script | None:
    row = get_script(db, script_id)
    if not row:
        return None
    if payload.name != row.name and not symbol_exists_on_nse(payload.name):
        raise ValueError("SCRIPT_NOT_FOUND")
    row.name = payload.name
    row.up_percent = _d(payload.up_percent)
    row.down_percent = _d(payload.down_percent)
    row.total_invest_amount = _d(payload.total_invest_amount)
    row.invest_slots = payload.invest_slots
    row.max_buy = payload.max_buy
    row.active = payload.active
    db.commit()
    db.refresh(row)
    return row


def delete_script(db: Session, script_id: int) -> bool:
    row = get_script(db, script_id)
    if not row:
        return False
    db.delete(row)
    db.commit()
    return True
